File: src/main/java/me/taubsie/dungeonhub/application/command/commands/CarryTierCommand.java
package me.taubsie.dungeonhub.application.command.commands;

import me.taubsie.dungeonhub.application.command.Command;
import me.taubsie.dungeonhub.application.command.CommandParameters;
import me.taubsie.dungeonhub.application.connection.DungeonHubConnection;
import me.taubsie.dungeonhub.application.connection.dungeon_hub.CarryTierConnection;
import me.taubsie.dungeonhub.application.connection.dungeon_hub.CarryTypeConnection;
import me.taubsie.dungeonhub.application.connection.dungeon_hub.DiscordServerConnection;
import me.taubsie.dungeonhub.kord.application.exceptions.CommandExecutionException;
import me.taubsie.dungeonhub.kord.application.exceptions.InvalidOptionException;
import me.taubsie.dungeonhub.kord.application.exceptions.InvalidSubCommandException;
import me.taubsie.dungeonhub.application.service.ApplicationService;
import me.taubsie.dungeonhub.common.model.carry_tier.CarryTierCreationModel;
import me.taubsie.dungeonhub.common.model.carry_tier.CarryTierModel;
import me.taubsie.dungeonhub.common.model.carry_tier.CarryTierUpdateModel;
import me.taubsie.dungeonhub.common.model.carry_type.CarryTypeModel;
import org.javacord.api.entity.DiscordEntity;
import org.javacord.api.entity.channel.Channel;
import org.javacord.api.entity.channel.ChannelCategory;
import org.javacord.api.entity.channel.ChannelType;
import org.javacord.api.entity.channel.ServerChannel;
import org.javacord.api.entity.permission.PermissionType;
import org.javacord.api.entity.server.Server;
import org.javacord.api.event.interaction.SlashCommandCreateEvent;
import org.javacord.api.interaction.SlashCommandInteractionOption;
import org.javacord.api.interaction.SlashCommandOption;
import org.javacord.api.interaction.SlashCommandOptionBuilder;
import org.javacord.api.interaction.SlashCommandOptionType;

import java.util.List;
import java.util.Optional;

@CommandParameters(name = "carry-tier", description = "Set up the carry tiers for this server.",
        enabledForPermissions = PermissionType.ADMINISTRATOR)
public class CarryTierCommand extends Command {
    public static final String FIELD_NAME = "carry-tier";

    public static SlashCommandOption getCarryTierOption() {
        return new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.STRING)
                .setName(FIELD_NAME)
                .setDescription("The identifier of the carry tier")
                .setRequired(true)
                .setMaxLength(30)
                .setAutocompletable(true)
                .build();
    }

    protected void executeCommand(SlashCommandCreateEvent slashCommandCreateEvent) {
        SlashCommandInteractionOption subCommand = getOptionAtIndex(0);

        switch (subCommand.getName().toLowerCase()) {
            case "create" -> create(subCommand);
            case "delete" -> delete(subCommand);
            case "get" -> get(subCommand);
            case "edit" -> edit(subCommand);
            case "reset" -> reset(subCommand);
            default -> throw new InvalidSubCommandException();
        }
    }

    public void create(SlashCommandInteractionOption subCommand) {
        Server server = getServer();

        Optional<CarryTypeModel> carryType = CarryTypeConnection.getInstance(server.getId())
                .getByIdentifier(getStringOption(subCommand, CarryTypeCommand.FIELD_NAME));

        if (carryType.isEmpty()) {
            throw new InvalidOptionException(CarryTypeCommand.FIELD_NAME, "Carry Type couldn't be found.");
        }

        String identifier = getStringOption(subCommand, "identifier")
                .strip()
                .toLowerCase()
                .replace(" ", "_");
        String displayName = getStringOption(subCommand, "display-name");

        if(CarryTierConnection.getInstance(carryType.get()).getByIdentifier(identifier).isPresent()) {
            throw new InvalidOptionException("identifier", "That carry tier already exists!");
        }

        CarryTierCreationModel creationModel = new CarryTierCreationModel();
        creationModel.setIdentifier(identifier);
        creationModel.setDisplayName(displayName);

        getOptionalStringOption(subCommand, "descriptive-name")
                .ifPresent(creationModel::setDescriptiveName);
        getOptionalChannelOption(subCommand, "category")
                .flatMap(Channel::asChannelCategory)
                .filter(channelCategory -> DiscordServerConnection.getInstance()
                        .getCarryTierFromCategory(server.getId(), channelCategory.getId()).isEmpty())
                .map(DiscordEntity::getId)
                .ifPresent(creationModel::setCategory);
        getOptionalChannelOption(subCommand, "price-channel")
                .map(DiscordEntity::getId)
                .ifPresent(creationModel::setPriceChannel);
        getOptionalStringOption(subCommand, "thumbnail-url")
                .ifPresent(creationModel::setThumbnailUrl);
        getOptionalStringOption(subCommand, "price-title")
                .ifPresent(creationModel::setPriceTitle);
        getOptionalStringOption(subCommand, "price-description")
                .ifPresent(creationModel::setPriceDescription);

        Optional<CarryTierModel> carryTier = CarryTierConnection.getInstance(carryType.get())
                .createCarryTier(creationModel);

        if (carryTier.isEmpty()) {
            //TODO custom class?
            throw new CommandExecutionException("Couldn't add that carry tier.");
        }

        respond(ApplicationService.getInstance()
                .getCarryTierEmbed(carryTier.get())
                .setTitle("Carry Tier created"));
    }

    public void delete(SlashCommandInteractionOption subCommand) {
        String carryTypeIdentifier = getStringOption(subCommand, CarryTypeCommand.FIELD_NAME);
        String identifier = getStringOption(subCommand, FIELD_NAME);

        Optional<CarryTypeModel> carryType = CarryTypeConnection.getInstance(getServer().getId())
                .getByIdentifier(carryTypeIdentifier);

        if (carryType.isEmpty()) {
            //TODO custom class
            throw new CommandExecutionException("That carry type doesn't exists!");
        }

        Optional<CarryTierModel> carryTier = CarryTierConnection.getInstance(carryType.get())
                .getByIdentifier(identifier);

        if (carryTier.isEmpty()) {
            throw new InvalidOptionException(FIELD_NAME);
        }

        if (!carryTier.get().getCarryType().equals(carryType.get())) {
            //TODO custom class
            throw new CommandExecutionException("Well this is weird.. Something doesn't really add up!");
        }

        Optional<CarryTierModel> deletedCarryTier = DungeonHubConnection.getInstance().removeCarryTier(carryTier.get());

        if (deletedCarryTier.isEmpty()) {
            //TODO custom class
            throw new CommandExecutionException();
        }

        respond(ApplicationService.getInstance()
                .getCarryTierEmbed(deletedCarryTier.get())
                .setTitle("Deleted Carry Tier"));
    }

    public void get(SlashCommandInteractionOption subCommand) {
        Optional<CarryTypeModel> carryType = CarryTypeConnection.getInstance(getServer().getId())
                .getByIdentifier(getStringOption(subCommand, CarryTypeCommand.FIELD_NAME));

        if (carryType.isEmpty()) {
            //TODO custom exception class
            throw new CommandExecutionException("Carry type not found.");
        }

        Optional<CarryTierModel> carryTier = CarryTierConnection.getInstance(carryType.get())
                .getByIdentifier(getStringOption(subCommand, FIELD_NAME));

        if (carryTier.isEmpty()) {
            throw new InvalidOptionException(FIELD_NAME, "That carry tier doesn't exist!");
        }

        respondEphemeral(ApplicationService.getInstance().getCarryTierEmbed(carryTier.get()));
    }

    public void edit(SlashCommandInteractionOption subCommand) {
        Server server = getServer();

        Optional<CarryTypeModel> carryType = CarryTypeConnection.getInstance(server.getId())
                .getByIdentifier(getStringOption(subCommand, CarryTypeCommand.FIELD_NAME));

        if (carryType.isEmpty()) {
            //TODO custom class
            throw new CommandExecutionException("That carry type doesn't exists!");
        }

        Optional<CarryTierModel> carryTier = CarryTierConnection.getInstance(carryType.get())
                .getByIdentifier(getStringOption(subCommand, FIELD_NAME));

        if (carryTier.isEmpty()) {
            throw new InvalidOptionException(FIELD_NAME, "That carry tier doesn't exist");
        }

        Optional<String> displayName = getOptionalStringOption(subCommand, "display-name");
        Optional<ChannelCategory> category = getOptionalChannelOption(subCommand, "category")
                .flatMap(Channel::asChannelCategory);
        Optional<ServerChannel> priceChannel = getOptionalChannelOption(subCommand, "price-channel");
        Optional<String> descriptiveName = getOptionalStringOption(subCommand, "descriptive-name");
        Optional<String> thumbnailUrl = getOptionalStringOption(subCommand, "thumbnail-url");
        Optional<String> priceTitle = getOptionalStringOption(subCommand, "price-title");

        if (displayName.isEmpty() && category.isEmpty() && priceChannel.isEmpty() && descriptiveName.isEmpty() && thumbnailUrl.isEmpty() && priceTitle.isEmpty()) {
            //TODO custom class
            throw new CommandExecutionException("Please provide something you want to edit.");
        }

        if (category.isPresent()) {
            Optional<CarryTierModel> categoryCarryTier =
                    DiscordServerConnection.getInstance()
                            .getCarryTierFromCategory(server.getId(), category.get().getId());
            if (categoryCarryTier.isPresent()) {
                respondEphemeral(ApplicationService.getInstance()
                        .getErrorEmbed(ApplicationService.getInstance().getCarryTierEmbed(categoryCarryTier.get()))
                        .setTitle("Carry Tier for that category is already present!"));
                return;
            }
        }

        CarryTierUpdateModel updateModel = CarryTierUpdateModel.fromCarryTier(carryTier.get());

        displayName.ifPresent(updateModel::setDisplayName);
        category.map(DiscordEntity::getId).ifPresent(updateModel::setCategory);
        priceChannel.map(DiscordEntity::getId).ifPresent(updateModel::setPriceChannel);
        descriptiveName.ifPresent(updateModel::setDescriptiveName);
        thumbnailUrl.ifPresent(updateModel::setThumbnailUrl);
        priceTitle.ifPresent(updateModel::setPriceTitle);

        Optional<CarryTierModel> updatedCarryTier = CarryTierConnection.getInstance(carryType.get())
                .updateCarryTier(carryTier.get().getId(), updateModel);

        if (updatedCarryTier.isEmpty()) {
            //TODO custom class
            throw new CommandExecutionException("Couldn't update carry tier.");
        }

        respond(ApplicationService.getInstance()
                .getCarryTierEmbed(updatedCarryTier.get())
                .setTitle("Updated Carry Tier"));
    }

    public void reset(SlashCommandInteractionOption subCommand) {
        Optional<CarryTypeModel> carryType = CarryTypeConnection.getInstance(getServer().getId())
                .getByIdentifier(getStringOption(subCommand, CarryTypeCommand.FIELD_NAME));

        if (carryType.isEmpty()) {
            //TODO custom class
            throw new CommandExecutionException("That carry type doesn't exists!");
        }

        Optional<CarryTierModel> carryTier = CarryTierConnection.getInstance(carryType.get())
                .getByIdentifier(getStringOption(subCommand, FIELD_NAME));

        if (carryTier.isEmpty()) {
            throw new InvalidOptionException(FIELD_NAME, "Carry tier doesn't exist");
        }

        Boolean category = getBooleanOption(subCommand, "category");
        Boolean priceChannel = getBooleanOption(subCommand, "price-channel");
        Boolean descriptiveName = getBooleanOption(subCommand, "descriptive-name");
        Boolean thumbnailUrl = getBooleanOption(subCommand, "thumbnail-url");
        Boolean priceTitle = getBooleanOption(subCommand, "price-title");

        if (!category && !priceChannel && !descriptiveName && !thumbnailUrl && !priceTitle) {
            //TODO custom class
            throw new CommandExecutionException("Please provide something you want to reset.");
        }

        CarryTierUpdateModel updateModel = CarryTierUpdateModel.fromCarryTier(carryTier.get());

        if (Boolean.TRUE.equals(category)) {
            updateModel.setCategory(-1L);
        }

        if (Boolean.TRUE.equals(priceChannel)) {
            updateModel.setPriceChannel(-1L);
        }

        if (Boolean.TRUE.equals(descriptiveName)) {
            updateModel.setDescriptiveName(null);
        }

        if (Boolean.TRUE.equals(thumbnailUrl)) {
            updateModel.setThumbnailUrl(null);
        }

        if (Boolean.TRUE.equals(priceTitle)) {
            updateModel.setPriceTitle(null);
        }

        Optional<CarryTierModel> updatedCarryTier = CarryTierConnection.getInstance(carryType.get())
                .updateCarryTier(carryTier.get().getId(), updateModel);

        if (updatedCarryTier.isEmpty()) {
            //TODO custom class
            throw new CommandExecutionException("Couldn't update carry tier.");
        }

        respond(ApplicationService.getInstance()
                .getCarryTierEmbed(updatedCarryTier.get())
                .setTitle("Updated Carry Tier with reset values"));
    }

    public List<SlashCommandOption> getSlashCommandOptions() {
        return List.of(getCreateCommand(), getDeleteCommand(), getGetCommand(), getEditCommand(), getResetCommand());
    }

    private SlashCommandOption getGetCommand() {
        return new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.SUB_COMMAND)
                .setName("get")
                .setDescription("Get information about a carry tier")
                .setOptions(List.of(CarryTypeCommand.getCarryTypeOption(), getCarryTierOption()))
                .build();
    }

    private SlashCommandOption getEditCommand() {
        SlashCommandOption displayNameOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.STRING)
                .setName("display-name")
                .setDescription("Set the display name of the carry tier")
                .build();

        SlashCommandOption descriptiveNameOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.STRING)
                .setName("descriptive-name")
                .setDescription("Set the descriptive name which replaces the display name in some places")
                .build();

        SlashCommandOption categoryOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.CHANNEL)
                .setChannelTypes(List.of(ChannelType.CHANNEL_CATEGORY))
                .setName("category")
                .setDescription("Set the category of the tickets")
                .build();

        SlashCommandOption priceChannelOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.CHANNEL)
                .setChannelTypes(List.of(ChannelType.SERVER_TEXT_CHANNEL))
                .setName("price-channel")
                .setDescription("Set the channel where the price list should appear")
                .build();

        SlashCommandOption thumbnailUrlOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.STRING)
                .setName("thumbnail-url")
                .setDescription("Set the thumbnail which is used to make some embeds look nicer")
                .build();

        SlashCommandOption priceTitleOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.STRING)
                .setName("price-title")
                .setDescription("Set the title of the price embed")
                .build();

        return new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.SUB_COMMAND)
                .setName("edit")
                .setDescription("Edit a carry tier")
                .setOptions(List.of(CarryTypeCommand.getCarryTypeOption(),
                        getCarryTierOption(),
                        displayNameOption,
                        descriptiveNameOption,
                        categoryOption,
                        priceChannelOption,
                        thumbnailUrlOption,
                        priceTitleOption))
                .build();
    }

    private SlashCommandOption getResetCommand() {
        SlashCommandOption descriptiveNameOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.BOOLEAN)
                .setName("descriptive-name")
                .setDescription("Reset the descriptive name which replaces the display name in some places")
                .setRequired(true)
                .build();

        SlashCommandOption categoryOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.BOOLEAN)
                .setName("category")
                .setDescription("Reset the category of the tickets")
                .setRequired(true)
                .build();

        SlashCommandOption priceChannelOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.BOOLEAN)
                .setName("price-channel")
                .setDescription("Reset the channel where the price list should appear")
                .setRequired(true)
                .build();

        SlashCommandOption thumbnailUrlOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.BOOLEAN)
                .setName("thumbnail-url")
                .setDescription("Reset the thumbnail which is used to make some embeds look nicer")
                .setRequired(true)
                .build();

        SlashCommandOption priceTitleOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.BOOLEAN)
                .setName("price-title")
                .setDescription("Reset the title of the price embed")
                .setRequired(true)
                .build();

        return new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.SUB_COMMAND)
                .setName("reset")
                .setDescription("Reset a carry tier")
                .setOptions(List.of(CarryTypeCommand.getCarryTypeOption(), getCarryTierOption(),
                        descriptiveNameOption, categoryOption, priceChannelOption, thumbnailUrlOption,
                        priceTitleOption))
                .build();
    }

    private SlashCommandOption getDeleteCommand() {
        return new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.SUB_COMMAND)
                .setName("delete")
                .setDescription("Delete a carry tier")
                .setOptions(List.of(CarryTypeCommand.getCarryTypeOption(), getCarryTierOption()))
                .build();
    }

    private SlashCommandOption getCreateCommand() {
        SlashCommandOption identifierOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.STRING)
                .setName("identifier")
                .setDescription("The identifier of the carry tier")
                .setRequired(true)
                .build();

        SlashCommandOption displayNameOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.STRING)
                .setName("display-name")
                .setDescription("The display name of the carry tier")
                .setRequired(true)
                .setMaxLength(30)
                .build();

        SlashCommandOption descriptiveNameOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.STRING)
                .setName("descriptive-name")
                .setDescription("Set the descriptive name which replaces the display name in some places")
                .build();

        SlashCommandOption categoryOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.CHANNEL)
                .setChannelTypes(List.of(ChannelType.CHANNEL_CATEGORY))
                .setName("category")
                .setDescription("Set the category of the tickets")
                .build();

        SlashCommandOption priceChannelOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.CHANNEL)
                .setChannelTypes(List.of(ChannelType.SERVER_TEXT_CHANNEL))
                .setName("price-channel")
                .setDescription("Set the channel where the price list should appear")
                .build();

        SlashCommandOption priceDescriptionOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.STRING)
                .setName("price-description")
                .setDescription("Set the price description which is shown on the top of the price message.")
                .build();

        SlashCommandOption thumbnailUrlOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.STRING)
                .setName("thumbnail-url")
                .setDescription("Set the thumbnail which is used to make some embeds look nicer")
                .build();

        SlashCommandOption priceTitleOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.STRING)
                .setName("price-title")
                .setDescription("Set the title of the price embed")
                .build();

        return new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.SUB_COMMAND)
                .setName("create")
                .setDescription("Create a new carry tier")
                .setOptions(List.of(CarryTypeCommand.getCarryTypeOption(), identifierOption, displayNameOption,
                        descriptiveNameOption, categoryOption, priceChannelOption, priceDescriptionOption,
                        thumbnailUrlOption, priceTitleOption))
                .build();
    }
}

File: src/main/kotlin/me/taubsie/dungeonhub/application/commands/CarryTierCommand.kt
package me.taubsie.dungeonhub.application.commands

import dev.kord.common.entity.ChannelType
import dev.kord.common.entity.Permission
import dev.kord.common.entity.Permissions
import dev.kordex.core.commands.Arguments
import dev.kordex.core.commands.application.slash.publicSubCommand
import dev.kordex.core.commands.converters.impl.boolean
import dev.kordex.core.commands.converters.impl.optionalChannel
import dev.kordex.core.commands.converters.impl.optionalString
import dev.kordex.core.commands.converters.impl.string
import dev.kordex.core.extensions.Extension
import dev.kordex.core.extensions.publicSlashCommand
import dev.kordex.core.i18n.toKey
import me.taubsie.dungeonhub.application.exceptions.CommandExecutionWarning
import me.taubsie.dungeonhub.application.exceptions.InvalidOptionException
import me.taubsie.dungeonhub.application.loader.LoadExtension
import me.taubsie.dungeonhub.application.service.ApplicationService
import me.taubsie.dungeonhub.application.service.AutoCompletionService
import net.dungeonhub.connection.CarryTierConnection
import net.dungeonhub.connection.CarryTypeConnection
import net.dungeonhub.connection.DiscordServerConnection
import net.dungeonhub.model.carry_tier.CarryTierCreationModel
import java.util.*

/**
 * This extension provides the slash commands for the carry tier management.
 * It allows the creation, deletion, editing and resetting of carry tiers.
 * The commands are only available in guilds and require the administrator permission.
 * The commands are:
 * - `/carry-tier create`
 * - `/carry-tier delete`
 * - `/carry-tier get`
 * - `/carry-tier edit`
 * - `/carry-tier reset`
 *
 * These commands are used to create, delete, get, edit and reset carry tiers.
 */
@LoadExtension
class CarryTierCommand : Extension() {
    override val name = "carry-tier-command"

    override suspend fun setup() {
        publicSlashCommand {
            name = "carry-tier".toKey()
            description = "Set up the carry tiers for this server.".toKey()
            defaultMemberPermissions = Permissions(Permission.Administrator)
            allowInDms = false

            publicSubCommand(::CarryTierCreateArguments) {
                name = "create".toKey()
                description = "Create a new carry tier".toKey()

                action {
                    respond {
                        val carryType =
                            CarryTypeConnection[guild!!.id.value.toLong()].getByIdentifier(arguments.carryType)
                                ?: throw InvalidOptionException("carry-type", "Carry Type couldn't be found.")

                        val identifier: String = arguments.identifier
                            .trim()
                            .lowercase(Locale.getDefault())
                            .replace(" ", "_")

                        if (CarryTierConnection[carryType].getByIdentifier(identifier) != null) {
                            throw InvalidOptionException("identifier", "That carry tier already exists!")
                        }

                        if (arguments.category != null && DiscordServerConnection.getCarryTierFromCategory(
                                guild!!.id.value.toLong(),
                                arguments.category!!.id.value.toLong()
                            ) != null
                        ) {
                            throw InvalidOptionException(
                                "category",
                                "That category is already assigned to another carry tier!"
                            )
                        }

                        val creationModel = CarryTierCreationModel(
                            identifier = identifier,
                            displayName = arguments.displayName,
                            category = arguments.category!!.id.value.toLong(),
                            priceChannel = arguments.priceChannel!!.id.value.toLong(),
                            descriptiveName = arguments.descriptiveName,
                            thumbnailUrl = arguments.thumbnailUrl,
                            priceTitle = arguments.priceTitle,
                            priceDescription = arguments.priceDescription
                        )

                        val carryTier = CarryTierConnection[carryType].createCarryTier(creationModel)
                            ?: throw CommandExecutionWarning("Couldn't add that carry tier.")

                        val embed = ApplicationService.getCarryTierEmbed(carryTier)
                        embed.title = "Carry Tier created"
                        embeds = mutableListOf(embed)
                    }
                }
            }

            publicSubCommand(::CarryTierArguments) {
                name = "delete".toKey()
                description = "Delete a carry tier".toKey()

                action {
                    respond {
                        val carryType =
                            CarryTypeConnection[guild!!.id.value.toLong()].getByIdentifier(arguments.carryType)
                                ?: throw CommandExecutionWarning("That carry type doesn't exists!")

                        val carryTier = CarryTierConnection[carryType].getByIdentifier(arguments.carryTier)
                            ?: throw InvalidOptionException("carry-tier")

                        if (carryTier.carryType != carryType) {
                            throw CommandExecutionWarning("Well this is weird.. Something doesn't really add up!")
                        }

                        val deletedCarryTier = CarryTierConnection[carryType].deleteCarryTier(carryTier.id)
                            ?: throw CommandExecutionWarning("Couldn't delete the carry tier.")

                        val embed = ApplicationService.getCarryTierEmbed(deletedCarryTier)
                        embed.title = "Deleted Carry Tier"
                        embeds = mutableListOf(embed)
                    }
                }
            }

            publicSubCommand(::CarryTierArguments) {
                name = "get".toKey()
                description = "Get information about a carry tier".toKey()

                action {
                    respond {
                        val carryType =
                            CarryTypeConnection[guild!!.id.value.toLong()].getByIdentifier(arguments.carryType)
                                ?: throw CommandExecutionWarning("Carry type not found.")

                        val carryTier = CarryTierConnection[carryType].getByIdentifier(arguments.carryTier)
                            ?: throw InvalidOptionException("carry-tier", "That carry tier doesn't exist!")

                        val embed = ApplicationService.getCarryTierEmbed(carryTier)
                        embeds = mutableListOf(embed)
                    }
                }
            }

            publicSubCommand(::CarryTierEditArguments) {
                name = "edit".toKey()
                description = "Edit a carry tier".toKey()

                action {
                    respond {
                        val carryType =
                            CarryTypeConnection[guild!!.id.value.toLong()].getByIdentifier(arguments.carryType)
                                ?: throw CommandExecutionWarning("That carry type doesn't exists!")

                        val carryTier = CarryTierConnection[carryType].getByIdentifier(arguments.carryTier)
                            ?: throw InvalidOptionException("carry-tier", "That carry tier doesn't exist")

                        if (arguments.displayName == null && arguments.category == null && arguments.priceChannel == null && arguments.descriptiveName == null && arguments.thumbnailUrl == null && arguments.priceTitle == null) {
                            throw CommandExecutionWarning("Please provide something you want to edit.")
                        }

                        if (arguments.category != null) {
                            val categoryCarryTier = DiscordServerConnection.getCarryTierFromCategory(
                                guild!!.id.value.toLong(),
                                arguments.category!!.id.value.toLong()
                            )

                            if (categoryCarryTier != null && categoryCarryTier != carryTier) {
                                val embed = ApplicationService.getErrorEmbed(
                                    ApplicationService.getCarryTierEmbed(categoryCarryTier)
                                )
                                embed.title = "Carry Tier for that category is already present!"
                                embeds = mutableListOf(embed)
                                return@respond
                            }
                        }

                        val updateModel = carryTier.getUpdateModel()

                        if (arguments.displayName != null) {
                            updateModel.displayName = arguments.displayName
                        }

                        if (arguments.category != null) {
                            updateModel.category = arguments.category!!.id.value.toLong()
                        }

                        if (arguments.priceChannel != null) {
                            updateModel.priceChannel = arguments.priceChannel!!.id.value.toLong()
                        }

                        if (arguments.descriptiveName != null) {
                            updateModel.descriptiveName = arguments.descriptiveName
                        }

                        if (arguments.thumbnailUrl != null) {
                            updateModel.thumbnailUrl = arguments.thumbnailUrl
                        }

                        if (arguments.priceTitle != null) {
                            updateModel.priceTitle = arguments.priceTitle
                        }

                        val updatedCarryTier = CarryTierConnection[carryType].updateCarryTier(carryTier.id, updateModel)
                            ?: throw CommandExecutionWarning("Couldn't update carry tier.")

                        val embed = ApplicationService.getCarryTierEmbed(updatedCarryTier)
                        embed.title = "Updated Carry Tier"
                        embeds = mutableListOf(embed)
                    }
                }
            }

            publicSubCommand(::CarryTierResetArguments) {
                name = "reset".toKey()
                description = "Reset a carry tier".toKey()

                action {
                    respond {
                        val carryType =
                            CarryTypeConnection[guild!!.id.value.toLong()].getByIdentifier(arguments.carryType)
                                ?: throw CommandExecutionWarning("That carry type doesn't exists!")

                        val carryTier = CarryTierConnection[carryType].getByIdentifier(arguments.carryTier)
                            ?: throw InvalidOptionException("carry-tier", "Carry tier doesn't exist")

                        if (!arguments.category && !arguments.priceChannel && !arguments.descriptiveName && !arguments.thumbnailUrl && !arguments.priceTitle) {
                            throw CommandExecutionWarning("Please provide something you want to reset.")
                        }

                        val updateModel = carryTier.getUpdateModel()

                        if (arguments.category) {
                            updateModel.category = null
                        }

                        if (arguments.priceChannel) {
                            updateModel.priceChannel = null
                        }

                        if (arguments.descriptiveName) {
                            updateModel.descriptiveName = null
                        }

                        if (arguments.thumbnailUrl) {
                            updateModel.thumbnailUrl = null
                        }

                        if (arguments.priceTitle) {
                            updateModel.priceTitle = null
                        }

                        val updatedCarryTier =
                            CarryTierConnection[carryType].updateCarryTier(carryTier.id, updateModel)
                                ?: throw CommandExecutionWarning("Couldn't update carry tier.")

                        val embed = ApplicationService.getCarryTierEmbed(updatedCarryTier)
                        embed.title = "Updated Carry Tier with reset values"
                        embeds = mutableListOf(embed)
                    }
                }
            }
        }
    }

    inner class CarryTierCreateArguments : Arguments() {
        val carryType by string {
            name = "carry-type".toKey()
            description = "The identifier of the carry type".toKey()
            maxLength = 30
            autoCompleteCallback = AutoCompletionService.carryType
        }

        val identifier by string {
            name = "identifier".toKey()
            description = "The identifier of the carry tier".toKey()
        }

        val displayName by string {
            name = "display-name".toKey()
            description = "The display name of the carry tier".toKey()
            maxLength = 30
        }

        val descriptiveName by optionalString {
            name = "descriptive-name".toKey()
            description = "Set the descriptive name which replaces the display name in some places".toKey()
        }

        val category by optionalChannel {
            name = "category".toKey()
            description = "Set the category of the tickets".toKey()
            requiredChannelTypes = mutableSetOf(ChannelType.GuildCategory)
        }

        val priceChannel by optionalChannel {
            name = "price-channel".toKey()
            description = "Set the channel where the price list should appear".toKey()
            requiredChannelTypes = mutableSetOf(ChannelType.GuildText)
        }

        val priceDescription by optionalString {
            name = "price-description".toKey()
            description = "Set the price description which is shown on the top of the price message.".toKey()
        }

        val thumbnailUrl by optionalString {
            name = "thumbnail-url".toKey()
            description = "Set the thumbnail which is used to make some embeds look nicer".toKey()
        }

        val priceTitle by optionalString {
            name = "price-title".toKey()
            description = "Set the title of the price embed".toKey()
        }
    }

    inner class CarryTierArguments : Arguments() {
        val carryType by string {
            name = "carry-type".toKey()
            description = "The identifier of the carry type".toKey()
            maxLength = 30
            autoCompleteCallback = AutoCompletionService.carryType
        }

        val carryTier by string {
            name = "carry-tier".toKey()
            description = "The identifier of the carry tier".toKey()
            maxLength = 30
            autoCompleteCallback = AutoCompletionService.carryTier
        }
    }

    inner class CarryTierEditArguments : Arguments() {
        val carryType by string {
            name = "carry-type".toKey()
            description = "The identifier of the carry type".toKey()
            maxLength = 30
            autoCompleteCallback = AutoCompletionService.carryType
        }

        val carryTier by string {
            name = "carry-tier".toKey()
            description = "The identifier of the carry tier".toKey()
            maxLength = 30
            autoCompleteCallback = AutoCompletionService.carryTier
        }

        val displayName by optionalString {
            name = "display-name".toKey()
            description = "Set the display name of the carry tier".toKey()
        }

        val descriptiveName by optionalString {
            name = "descriptive-name".toKey()
            description = "Set the descriptive name which replaces the display name in some places".toKey()
        }

        val category by optionalChannel {
            name = "category".toKey()
            description = "Set the category of the tickets".toKey()
            requiredChannelTypes = mutableSetOf(ChannelType.GuildCategory)
        }

        val priceChannel by optionalChannel {
            name = "price-channel".toKey()
            description = "Set the channel where the price list should appear".toKey()
            requiredChannelTypes = mutableSetOf(ChannelType.GuildText)
        }

        val thumbnailUrl by optionalString {
            name = "thumbnail-url".toKey()
            description = "Set the thumbnail which is used to make some embeds look nicer".toKey()
        }

        val priceTitle by optionalString {
            name = "price-title".toKey()
            description = "Set the title of the price embed".toKey()
        }
    }

    inner class CarryTierResetArguments : Arguments() {
        val carryType by string {
            name = "carry-type".toKey()
            description = "The identifier of the carry type".toKey()
            maxLength = 30
            autoCompleteCallback = AutoCompletionService.carryType
        }

        val carryTier by string {
            name = "carry-tier".toKey()
            description = "The identifier of the carry tier".toKey()
            maxLength = 30
            autoCompleteCallback = AutoCompletionService.carryTier
        }

        val descriptiveName by boolean {
            name = "descriptive-name".toKey()
            description = "Reset the descriptive name which replaces the display name in some places".toKey()
        }

        val category by boolean {
            name = "category".toKey()
            description = "Reset the category of the tickets".toKey()
        }

        val priceChannel by boolean {
            name = "price-channel".toKey()
            description = "Reset the channel where the price list should appear".toKey()
        }

        val thumbnailUrl by boolean {
            name = "thumbnail-url".toKey()
            description = "Reset the thumbnail which is used to make some embeds look nicer".toKey()
        }

        val priceTitle by boolean {
            name = "price-title".toKey()
            description = "Reset the title of the price embed".toKey()
        }
    }
}